### PR TITLE
feat: add support to deeplink and file association on macOS

### PR DESCRIPTION
**Pull request by meowtec**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

This feature is necessary for the file-association function: open files with tauri-based app.
ref: https://github.com/tauri-apps/tauri/pull/4320

### CODE REVIEW

File: src/platform_impl/macos/app_delegate.rs [hunk 1]
```diff
@@ -92,3 +101,17 @@ extern "C" fn application_will_terminate(_: &Object, _: Sel, _: id) {
   AppState::exit();
   trace!("Completed `applicationWillTerminate`");
 }
+
+extern "C" fn application_open_file(_: &Object, _: Sel, _: id, file: id) -> i8 {
```
Comment (keiya01) [inline src/platform_impl/macos/app_delegate.rs]:
```suggestion
extern "C" fn application_open_file(_: &Object, _: Sel, _: id, file: id) -> bool {
```

File: src/platform_impl/macos/app_delegate.rs [hunk 2]
```diff
@@ -44,6 +49,10 @@ lazy_static! {
       sel!(applicationWillTerminate:),
       application_will_terminate as extern "C" fn(&Object, Sel, id),
     );
+    decl.add_method(
+      sel!(application:openFile:),
+      application_open_file as extern "C" fn(&Object, Sel, id, id) -> i8,
```
Comment (keiya01) [inline src/platform_impl/macos/app_delegate.rs]:
```suggestion
      application_open_file as extern "C" fn(&Object, Sel, id, id) -> bool,
```

**Review by keiya01 — CHANGES_REQUESTED** [review]
Thank you for your PR. Please check my review.

File: src/event.rs [hunk 3]
```diff
@@ -100,6 +100,11 @@ pub enum Event<'a, T: 'static> {
     position: PhysicalPosition<f64>,
   },
 
+  /// Emitted when open an external file with the app
+  /// ## Platform-specific
+  /// - **Windows / Android / Linux:** Unsupported.
```
Comment (keiya01) [inline src/event.rs]:
```suggestion
  /// Emitted when open an external file with the app
  ///
  /// ## Platform-specific
  ///
  /// - **Windows / Android / Linux:** Unsupported.
```

File: src/platform_impl/macos/app_delegate.rs (hunk 1 above)
Comment (meowtec) [reply src/platform_impl/macos/app_delegate.rs]: It seems should to be `BOOL`(`cocoa::base::BOOL`)

**Comment by FabianLars** [comment]
Did you try using <https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application?language=objc> instead?

I'm asking because that's what we need for "deep linking" support and it conflicts with openFile, and i'm not sure yet how we want to handle this conflict. I guess if we can use one interface for both that would be the easiest/best solution.

Anyway, i wanted to try making deep linking work on macos tomorrow, so i can try it out myself too.

**Comment by meowtec** [comment]
> Did you try using https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application?language=objc instead?
> 
> I'm asking because that's what we need for "deep linking" support and it conflicts with openFile, and i'm not sure yet how we want to handle this conflict. I guess if we can use one interface for both that would be the easiest/best solution.
> 
> Anyway, i wanted to try making deep linking work on macos tomorrow, so i can try it out myself too.

I did not aware of the conflict before, and i will try `openURLs` and find how to make both situations work.

**Comment by meowtec** [comment]
> Did you try using https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application?language=objc instead?
> 
> I'm asking because that's what we need for "deep linking" support and it conflicts with openFile, and i'm not sure yet how we want to handle this conflict. I guess if we can use one interface for both that would be the easiest/best solution.
> 
> Anyway, i wanted to try making deep linking work on macos tomorrow, so i can try it out myself too.

I have tried `openURLs` and it seems that we can use `openURLs` instead of `openFile` for handling file. When using `openURLs`, a `NSArray<NSURL>` are received, in which files are like `file:///path/to/file`

**Comment by lucasfernog** [comment]
I think `OpenFile` is the correct way to implement this. Let's implement deep link after this so we can see how to handle it.

Electron has both events btw.

**Comment by lucasfernog** [comment]
Actually I saw it gets weird on multiple files :( 

This scares me about openURLs `urls` argument though:
> An array of URLs to open. The list does not include URLs for which your app has a defined document type.

**Comment by amrbashir** [comment]
maybe we should add both events? `OpenFile` and `OpenUrl`? seems like both can be used in different cases and I don't see a harm in adding both

**Comment by amrbashir** [comment]
after a quick look at the documentation, it seems like it needs a return value to indicate the sucess/failure so maybe we need to do this in a similar way to this winit PR https://github.com/rust-windowing/winit/pull/1759

**Comment by FabianLars** [comment]
> maybe we should add both events? OpenFile and OpenUrl? seems like both can be used in different cases and I don't see a harm in adding both

If past-me is any trustworthy, we can't because openfile isn't triggered if an app listens to openurl

**Comment by amrbashir** [comment]
is there an resource confirming that? electron seems to have it

**Comment by FabianLars** [comment]
The docs i linked above mention this: <https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application?language=objc>

But these are apple's docs so it's likely that i'm misunderstanding them (or that they are wrong lol)

**Comment by amrbashir** [comment]
electron doesn't mention anything about this in their docs so I wonder how electron behaves if you register for both events, anyways, lets follow the official docs.

And in light of this limitation, we need to use the callback style then like this PR https://github.com/rust-windowing/winit/pull/1759 so it can be opt-in for whichever event you want.

**Comment by lucasfernog** [comment]
We could also emit OpenFile if the url is `file://` and OpenUrl for other urls. I couldn't find it in the electron codebase, maybe it's related to chromium.

**Comment by amrbashir** [comment]
maybe that's something that could be offloaded to tauri instead and tao just allows registering both and documents that only one should be registered.

**Comment by Toldoven** [comment]
What is the progress on this issue?

**Comment by amrbashir** [comment]
@wusyong could you check out this PR when you have some time? especially the `OpenFile` event and see if modifying the mutable reference `success` has any effect at all.

**Comment by wusyong** [comment]
@amrbashir How do I test this? Is there an example to run?

**Comment by amrbashir** [comment]
@wusyong 
```rs
use tao::{
  event::{Event, WindowEvent},
  event_loop::{ControlFlow, EventLoop},
  window::WindowBuilder,
};

#[allow(clippy::single_match)]
fn main() {
  let event_loop = EventLoop::new();

  let window = Some(
    WindowBuilder::new()
      .with_title("A fantastic window!")
      .with_inner_size(tao::dpi::LogicalSize::new(128.0, 128.0))
      .build(&event_loop)
      .unwrap(),
  );

  event_loop.run(move |event, _, control_flow| {
    *control_flow = ControlFlow::Wait;
    println!("{:?}", event);

    match event {
      Event::OpenFile { filename, success } => {
        dbg!(filename);
        *success = false; // reject the drop, change to true to accept it
      }
      Event::WindowEvent {
        event: WindowEvent::Destroyed,
        window_id: _,
        ..
      } => {
        *control_flow = ControlFlow::Exit;
      }
      Event::MainEventsCleared => {
        if let Some(w) = &window {
          w.request_redraw();
        }
      }
      _ => (),
    }
  });
}

```

**Comment by wusyong** [comment]
I tired example above but it couldn't really test if events work. If anyone knows how to test it, please let me know.

File: src/platform_impl/macos/app_delegate.rs [hunk 4]
```diff
@@ -96,3 +122,55 @@ extern "C" fn application_will_terminate(_: &Object, _: Sel, _: id) {
...
+  let filename = unsafe { CStr::from_ptr(file.UTF8String()) }
+    .to_string_lossy()
+    .to_string();
+  let mut success = true;
+
+  trace!("Get `application:openFile:` URLs: {:?}", filename);
+  AppState::open_file(filename, &mut success);
+  trace!("Completed `application:openFile:`");
+
+  success as i8
```
Comment (TrueCarry) [inline src/platform_impl/macos/app_delegate.rs]: This line throws error since return should be BOOL

**Comment by lucasfernog** [comment]
I'm testing it again and seems like `openURLs` is enough for our use case; The URL scheme is `file://` when opening files, and the deeplink scheme otherwise. Did I forget something @FabianLars ? This PR is so old I always forget something, but from my tests it looks enough and we could drop the openFile[s] logic.

**Comment by FabianLars** [comment]
I honestly forgot most of it too, but yeah, openurls is enough unless we need to respond to the openfile event (tho tbh i've not seen a usecase for that yet) and and could be used to simplify the implementation 🤷 

One thing tho, did you test it with multiple files too? we've had issues with that when we tried to abuse my deep link plugin for that.
Also testing the behavior of different plist configs, like only urls, only files, and having both. Again using the deep link plugin as an example, some users had to add the `file` scheme to the plist for some messed up reason.

So if all of that works i personally don't see much of a reason to keep openFile/s around

**Comment by lucasfernog** [comment]
I did test with multiple files, seems to work fine. I guess we can go with the openURLs approach and wait for community feedback.

**Comment by amrbashir** [comment]
I'd support openFile but default to openURLs because this will and should be upstreamed to winit and could serve as a reference the en

**Comment by lucasfernog** [comment]
@amrbashir when we upstream we could always use [`cfc768f` (#422)](https://github.com/tauri-apps/tao/pull/422/commits/cfc768f14af8dcd148bff2e5de83f4fdfa57f7d9) as a reference. But at least for our own use case it seems like openURLs is enough since it also support file opening.

**Comment by amrbashir** [comment]
yeah I know it is enough but I never want to restrict tao to an API and if possible allow both ways if it is not much hassle. Just to be safe from a future feature request where someone might request it.

**Review by amrbashir — APPROVED** [review]

**Pull request merged**